File: 백준/Platinum/1310. 달리기 코스/달리기 코스.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

    static class Point {
        int x;
        int y;

        Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        long result = 0;

        int n = Integer.parseInt(st.nextToken());
        List<Point> points = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            points.add(new Point(x, y));
        }
        points.sort((p1, p2) -> {
            if (p1.x == p2.x) return Integer.compare(p1.y, p2.y);
            return Integer.compare(p1.x, p2.x);
        });

        List<Point> convexHull = initConvexHull(points);

        int j = 1;
        Point standardPoint = new Point(0, 0); // 원점, 기준점이 되는 포인트
        for (int i = 0; i < convexHull.size(); i++) {
            int iNext = (i + 1) % convexHull.size();
            while (true) {
                int jNext = (j + 1) % convexHull.size();
                Point iVector = new Point(convexHull.get(iNext).x - convexHull.get(i).x,
                        convexHull.get(iNext).y - convexHull.get(i).y);
                Point jVector = new Point(convexHull.get(jNext).x - convexHull.get(j).x,
                        convexHull.get(jNext).y - convexHull.get(j).y);
                if (ccw(standardPoint, iVector, jVector) < 0) j = jNext;
                else break;
            }

            result = Math.max(result, getPowDistance(convexHull.get(i), convexHull.get(j)));
        }
        System.out.println(result);
    }

    // 반시계 방향으로 컨벡스헐 구성
    static List<Point> initConvexHull(List<Point> points) {
        List<Point> convexHull = new ArrayList<>();
        List<Point> lower = new ArrayList<>();
        List<Point> upper = new ArrayList<>();

        for (int i = 0 ; i < points.size() ; i++) {
            while (lower.size() >= 2 && ccw(lower.get(lower.size() - 2), lower.get(lower.size() - 1), points.get(i)) >= 0) {
                lower.remove(lower.size() - 1);
            }
            lower.add(points.get(i));
        }

        for (int i = points.size() - 1 ; i >= 0; i--) {
            while (upper.size() >= 2 && ccw(upper.get(upper.size() - 2), upper.get(upper.size() - 1), points.get(i)) >= 0) {
                upper.remove(upper.size() - 1);
            }
            upper.add(points.get(i));
        }

        lower.remove(lower.size() - 1);
        upper.remove(upper.size() - 1);

        convexHull.addAll(lower);
        convexHull.addAll(upper);
        return convexHull;
    }

    static long ccw(Point p1, Point p2, Point p3) {
        return (long) (p2.x - p1.x) * (p3.y - p2.y) - (long) (p3.x - p2.x) * (p2.y - p1.y);
    }

    static long getPowDistance(Point p1, Point p2) {
        return (long) (Math.pow(p2.x - p1.x, 2) + Math.pow(p2.y - p1.y, 2));
    }
}

File: 백준/Platinum/1310. 달리기 코스/달리기 코스.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

struct Point {
    int x, y;
    
    Point(int x = 0, int y = 0) : x(x), y(y) {}
};

// CCW (Counter Clockwise) 연산
long long ccw(Point p1, Point p2, Point p3) {
    return (long long)(p2.x - p1.x) * (p3.y - p2.y) - (long long)(p3.x - p2.x) * (p2.y - p1.y);
}

// 두 점 사이의 거리의 제곱을 계산
long long getPowDistance(Point p1, Point p2) {
    return (long long)(p2.x - p1.x) * (p2.x - p1.x) + (long long)(p2.y - p1.y) * (p2.y - p1.y);
}

// 컨벡스 헐 (Convex Hull) 계산
vector<Point> initConvexHull(vector<Point>& points) {
    vector<Point> lower, upper, convexHull;

    // x 좌표 기준으로 정렬, 동일하면 y 기준 정렬
    sort(points.begin(), points.end(), [](const Point& p1, const Point& p2) {
        return (p1.x == p2.x) ? (p1.y < p2.y) : (p1.x < p2.x);
    });

    for (size_t i = 0; i < points.size(); i++) {
        while (lower.size() >= 2 && ccw(lower[lower.size() - 2], lower[lower.size() - 1], points[i]) >= 0) {
            lower.pop_back();
        }
        lower.push_back(points[i]);
    }

    for (size_t i = points.size(); i-- > 0; ) {
        while (upper.size() >= 2 && ccw(upper[upper.size() - 2], upper[upper.size() - 1], points[i]) >= 0) {
            upper.pop_back();
        }
        upper.push_back(points[i]);
    }

    lower.pop_back();
    upper.pop_back();

    convexHull.insert(convexHull.end(), lower.begin(), lower.end());
    convexHull.insert(convexHull.end(), upper.begin(), upper.end());

    return convexHull;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;
    vector<Point> points(n);

    for (int i = 0; i < n; i++) {
        cin >> points[i].x >> points[i].y;
    }

    vector<Point> convexHull = initConvexHull(points);

    int j = 1;
    long long result = 0;
    Point standardPoint(0, 0);

    for (size_t i = 0; i < convexHull.size(); i++) {
        size_t iNext = (i + 1) % convexHull.size();
        while (true) {
            size_t jNext = (j + 1) % convexHull.size();
            Point iVector(convexHull[iNext].x - convexHull[i].x, convexHull[iNext].y - convexHull[i].y);
            Point jVector(convexHull[jNext].x - convexHull[j].x, convexHull[jNext].y - convexHull[j].y);

            if (ccw(standardPoint, iVector, jVector) < 0) {
                j = jNext;
            } else {
                break;
            }
        }
        result = max(result, getPowDistance(convexHull[i], convexHull[j]));
    }

    cout << result << "\n";
    return 0;
}
